pages/DSNP/Identity.md

```diff
@@ -26,6 +26,22 @@ the ability to delegate permission to others to announce content on the user's b
 * Announcements from a delegate MUST be able to be verified as to which delegate made the specific announcement.
 * Delegation revocation MUST NOT be retroactive.
 
+## Operations
+
+Each of the these MUST have verifiable approval of the acting
+principal(s) via signature proof or other system. For example, an
+implementation might use the public key of an asymmetric key pair as a
+control key, and provide a proof for each operation by producing a
+cryptographic signature of the inputs and some nonce value.
+
+| Operation | Optional? | Principal(s) | Inputs | Outputs/State Changes |
+|---------- |---------- |------------- |------- |-------------- |
+| Create DSNP Identifier | no | User | None | New DSNP Identifier connected to User's Control Key |
+| Create or Update Delegation | no | User AND Delegate | User's DSNP Identifier, Delegate's DSNP Identifier, Set of Announcement Permissions | Proof of Delegation |
```

Comment:
I know we talked about this some and I think I was on the combined side, but I am still on the fence about if this should be one or two lines. Thinking now about the `DSNP on [Impl]` side where it could be separate and I'm thinking now we should always have a 1:m relationship with implementations. Implementations could have more than one per item in the Ops list, but it would never have to split the ops list for operations that do just half of what is specified. Thoughts?

```suggestion
| Create Delegation | no | User AND Delegate | User's DSNP Identifier, Delegate's DSNP Identifier, Set of Announcement Permissions | Proof of Delegation |
| Update Delegation | no | User AND Delegate | User's DSNP Identifier, Delegate's DSNP Identifier, Set of Announcement Permissions | Proof of Delegation with New Permissions |
```

Reply:
From the point of view of an observable output record (the proof of delegation), I'm not sure there needs to be any distinction as long as the operation is idempotent. I see "Create or Update" (upsert, in developer-speak) a bit like a PUT operation in REST. Do you see a use case where an implementation would need to split it?

Reply:
Split use case is when an implementation needs to split it. Blockchain actions that could be combined are often separated to cause less on-chain work (and thus cost less). So I think it likely that an implementation have create and update separate.

It is different in Frequency:

- Create: `grant_delegation`
- Update: `grant_schema_permissions` AND `revoke_schema_permissions`

Reply:
Is there a way to specify create/update and also allow them to be a united operation so long as the behavior is as specified?

Reply:
Let's imagine a hypothetical "translation layer" converting DSNP operations to Frequency RPC calls. As long as this translation layer has the ability to access current state data (i.e. what delegation exists already), it can translate a "Define Delegation" call into any of grant_delegation, grant_schema_permissions, and revoke_schema_permissions. So for simplicity's sake I think there's a path to leaving it as a single operation in this spec.